compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs

```diff
@@ -298,53 +267,91 @@ getFctDocs ctx@DocCtx{..} cl_nameM (DeclData decl docs) = do
+        opMap = MS.fromList
+            [ (Fieldname $ packId id, (id, dmInfoM))
+            | (id, dmInfoM) <- classOpItems tycls ]
+        cl_methods = concatMap (getMethodDocs opMap cl_anchor cl_name cl_args) subDecls
         cl_super = do
-            tycon <- tyconMb
-            cls <- tyConClass_maybe tycon
-            let theta = classSCTheta cls
+            let theta = classSCTheta tycls
             guard (notNull theta)
             Just (TypeTuple $ map (typeToType ctx) theta)
         cl_instances = Nothing -- filled out later in 'distributeInstanceDocs'
     guard (exportsType dc_exports cl_name)
     Just ClassDoc {..}
   where
-    f :: Typename -> LSig GhcPs -> [FunctionDoc]
-    f cl_name (L dloc (ClassOpSig p b names n)) = catMaybes
-      [ getFctDocs ctx (Just cl_name) (DeclData (L dloc (SigD noExt (ClassOpSig p b [L loc name] n))) (MS.lookup name subdocs))
-      | L loc name <- names
-      ]
-    f _ _ = []
+    -- All documentation of class members is stored in tcdDocs.
+    -- To associate the docs with the correct member, we convert all members
+    -- and docs to declarations, sort them by their location
+    -- and then use collectDocs.
+    -- This is the equivalent of Haddock’s Haddock.Interface.Create.classDecls.
+    subDecls :: [(LHsDecl GhcPs, Maybe DocText)]
+    subDecls = collectDocs . sortOn getLoc $ decls
+    decls = docs ++ defs ++ sigs ++ ats
+    docs = map (fmap (DocD noExt)) tcdDocs
+    defs = map (fmap (ValD noExt)) (bagToList tcdMeths)
+    sigs = map (fmap (SigD noExt)) tcdSigs
+    ats = map (fmap (TyClD noExt . FamDecl noExt)) tcdATs
+
+    -- | Extract typeclass method docs from a subDecl. Notice that
+    -- we may have to generate multiple docs from a single declaration,
+    -- thanks to declarations of the form
+    --
+    -- @
+    --     (+), (-) : t -> t -> t
+    -- @
+    --
+    -- where multiple names are present. In that case we just duplicate
+    -- the associated docs for each name that is defined.
+    getMethodDocs ::
+        MS.Map Fieldname ClassOpItem
+        -> Maybe Anchor
+        -> Typename
+        -> [T.Text]
+        -> (LHsDecl GhcPs, Maybe DocText)
+        -> [ClassMethodDoc]
+    getMethodDocs opMap cl_anchor cl_name cl_args = \case
+        (L _ (SigD _ (ClassOpSig _ cm_isDefault rdrNamesL _)), cm_descr) ->
+            flip mapMaybe rdrNamesL $ \rdrNameL -> do
+                let cm_name = Fieldname . packRdrName . unLoc $ rdrNameL
+                    cm_anchor = guard (not cm_isDefault) >>
+                        Just (functionAnchor dc_modname cm_name)
+                (id, dmInfoM) <- MS.lookup cm_name opMap
+                ty <- case (cm_isDefault, dmInfoM) of
+                    (True, Nothing) ->
+                        error "getMethodDocs: expected default method to have associated default method info"
+                    (True, Just (_, GenericDM ty)) -> Just ty
+                    _ -> Just (idType id)
```

Comment:
Can it ever happen that we get `(False, Just …)`? If that should also be an error I would suggest to change this to only case on `dmInfoM` and instead have something like `when (cm_isDefault /= isJust dmInfoM) $ error …`.

Reply:
`(False, Just ...)` is expected whenever we're processing a typeclass method's original signature, but there also exists a default implementation. It's not an error. The ``

Reply:
Ah I see, in that case maybe spell that out in a comment? :)

Reply:
👍